### DISCUSSION TITLE
Asking about lower NVIDIA GPUs minimum version requirement for CUDA

### DISCUSSION THREAD

**Discussion by TAT-Hins** [immich-app/immich#13545] [Q&A] [answered]
Hi guys, thank you so much for your great jobs on immich that helps me a lot in managing tons of my photos.

I have a spare outdated Quadro K2200 and would like to make use of it, while its highest CUDA version is 5.0, lower than the minimum requirement of immich (CUDA 5.2). I would like to ask if there's any way to lower the minimum support version for CUDA in immich-machine-learning, such as customizing and building my own binary and docker images.

Again, sincerely appreciate your great work!

**Comment by mertalev** [comment, marked as answer]
You can try [building](https://github.com/microsoft/onnxruntime/blob/main/dockerfiles/Dockerfile.cuda) ONNX Runtime yourself and setting `CMAKE_CUDA_ARCHITECTURES` accordingly. Then extend the immich machine learning image, remove the onnxruntime-gpu package, copy over the one you built and install it.

**Reply by TAT-Hins** [reply]
> You can try [building](https://github.com/microsoft/onnxruntime/blob/main/dockerfiles/Dockerfile.cuda) ONNX Runtime yourself and setting `CMAKE_CUDA_ARCHITECTURES` accordingly. Then extend the immich machine learning image, remove the onnxruntime-gpu package, copy over the one you built and install it.

Thanks for the guidance. I'll give some time to learn about ONNX Runtime and see if I can give it a try to build a runtime with CUDA 5.0

**Reply by blackthorne** [reply]
@TAT-Hins this worked for me:
```FROM python:3.11-bookworm@sha256:fc1f2e357c307c4044133952b203e66a47e7726821a664f603a180a0c5823844 AS builder-cpu

FROM nvidia/cuda:11.8.0-cudnn8-devel-ubuntu22.04 AS builder-cuda

# Build ONNX Runtime wheel with CUDA support targeting compute capability 5.0 (Maxwell, e.g. GTX 960M)
RUN set -eux; \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
      git wget ca-certificates \
      python3.11 python3.11-venv python3.11-distutils python3.11-dev python3-pip \
      build-essential zlib1g-dev pkg-config; \
    rm -rf /var/lib/apt/lists/*

# Install a recent CMake
RUN wget -q https://github.com/Kitware/CMake/releases/download/v3.28.3/cmake-3.28.3-linux-x86_64.sh && \
    chmod +x cmake-3.28.3-linux-x86_64.sh && \
    ./cmake-3.28.3-linux-x86_64.sh --skip-license --prefix=/usr/local && \
    rm cmake-3.28.3-linux-x86_64.sh && \
    ln -sf /usr/local/bin/cmake /usr/bin/cmake

# Python deps required by ONNX Runtime build tooling (NumPy is required for Python bindings)
RUN python3.11 -m pip install --no-cache-dir --upgrade pip && \
    python3.11 -m pip install --no-cache-dir "numpy<2.0" psutil packaging

# Target CUDA 5.0
ENV CMAKE_CUDA_ARCHITECTURES=50
ENV CUDA_ARCHITECTURES=50

# Create version.json for ONNX Runtime build script (missing in CUDA 11.8 images)
RUN echo '{"cuda": {"name": "CUDA SDK", "version": "11.8.0"}}' > /usr/local/cuda/version.json

WORKDIR /workspace
# Clone ONNX Runtime (shallow for speed)
RUN git clone --recursive --depth 1 https://github.com/microsoft/onnxruntime.git onnxruntime

WORKDIR /workspace/onnxruntime
# Build ONNX Runtime with CUDA, disable contrib ops (bfloat16) to be Maxwell-safe, and build wheel
RUN python3.11 tools/ci_build/build.py \
    --build_dir /workspace/onnxruntime/build \
    --config Release \
    --parallel 8 \
    --build_shared_lib \
    --use_cuda \
    --cuda_home /usr/local/cuda \
    --cudnn_home /usr \
    --cmake_extra_defines CMAKE_CUDA_ARCHITECTURES=50 \
    --disable_contrib_ops \
    --skip_tests \
    --build_wheel \
    --allow_running_as_root

# Debug: Check where the wheel was created
RUN find /workspace/onnxruntime/build -name "*.whl" -type f

# Move the wheel to /opt
RUN find /workspace/onnxruntime/build -name "*.whl" -type f -exec mv {} /opt/ \;

FROM builder-cpu AS builder

ARG DEVICE
ENV PYTHONDONTWRITEBYTECODE=1 \
    PYTHONUNBUFFERED=1 \
    VIRTUAL_ENV=/opt/venv

RUN apt-get update && apt-get install -y --no-install-recommends g++

COPY --from=ghcr.io/astral-sh/uv:0.8.15@sha256:a5727064a0de127bdb7c9d3c1383f3a9ac307d9f2d8a391edc7896c54289ced0 /uv /uvx /bin/
RUN --mount=type=cache,target=/root/.cache/uv \
    --mount=type=bind,source=uv.lock,target=uv.lock \
    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
    uv sync --frozen --no-dev --no-editable --no-install-project --compile-bytecode --no-progress --active --link-mode copy

COPY --from=builder-cuda /opt/*.whl /opt/
RUN uv pip install /opt/*.whl    

FROM python:3.11-slim-bookworm@sha256:873f91540d53b36327ed4fb018c9669107a4e2a676719720edb4209c4b15d029 AS prod-cpu

ENV LD_PRELOAD=/usr/lib/libmimalloc.so.2 \
    MACHINE_LEARNING_MODEL_ARENA=false

FROM nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04 AS prod-cuda

ENV LD_PRELOAD=/usr/lib/libmimalloc.so.2 \
    MACHINE_LEARNING_MODEL_ARENA=false \
    NVIDIA_VISIBLE_DEVICES=all \
    NVIDIA_DRIVER_CAPABILITIES=compute,utility,video

# Install Python 3.11
RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
      python3.11 python3.11-venv python3-pip && \
    rm -rf /var/lib/apt/lists/* && \
    update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.11 1 && \
    update-alternatives --install /usr/bin/python python /usr/bin/python3.11 1

FROM prod-cuda AS prod

ARG DEVICE

RUN apt-get update && \
    apt-get install -y --no-install-recommends tini ccache libgl1 libglib2.0-0 libgomp1 $(if ! [ "$DEVICE" = "openvino" ] && ! [ "$DEVICE" = "rocm" ]; then echo "libmimalloc2.0"; fi) && \
    apt-get autoremove -yqq && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

RUN ln -s "/usr/lib/$(arch)-linux-gnu/libmimalloc.so.2" /usr/lib/libmimalloc.so.2

WORKDIR /usr/src
ENV TRANSFORMERS_CACHE=/cache \
    PYTHONDONTWRITEBYTECODE=1 \
    PYTHONUNBUFFERED=1 \
    PATH="/opt/venv/bin:$PATH" \
    PYTHONPATH=/opt/venv/lib/python3.11/site-packages:/usr/src \
    DEVICE=${DEVICE} \
    VIRTUAL_ENV=/opt/venv \
    MACHINE_LEARNING_CACHE_FOLDER=/cache \
    OMP_NUM_THREADS=8 \
    OMP_WAIT_POLICY=PASSIVE

# prevent core dumps
RUN echo "hard core 0" >> /etc/security/limits.conf && \
    echo "fs.suid_dumpable 0" >> /etc/sysctl.conf && \
    echo 'ulimit -S -c 0 > /dev/null 2>&1' >> /etc/profile

COPY --from=builder /opt/venv /opt/venv
COPY scripts/healthcheck.py .
COPY immich_ml immich_ml

ARG BUILD_ID
ARG BUILD_IMAGE
ARG BUILD_SOURCE_REF
ARG BUILD_SOURCE_COMMIT

ENV IMMICH_BUILD=${BUILD_ID}
ENV IMMICH_BUILD_URL=https://github.com/immich-app/immich/actions/runs/${BUILD_ID}
ENV IMMICH_BUILD_IMAGE=${BUILD_IMAGE}
ENV IMMICH_BUILD_IMAGE_URL=https://github.com/immich-app/immich/pkgs/container/immich-machine-learning
ENV IMMICH_REPOSITORY=immich-app/immich
ENV IMMICH_REPOSITORY_URL=https://github.com/immich-app/immich
ENV IMMICH_SOURCE_REF=${BUILD_SOURCE_REF}
ENV IMMICH_SOURCE_COMMIT=${BUILD_SOURCE_COMMIT}
ENV IMMICH_SOURCE_URL=https://github.com/immich-app/immich/commit/${BUILD_SOURCE_COMMIT}

ENTRYPOINT ["tini", "--"]
CMD ["python", "-m", "immich_ml"]

HEALTHCHECK CMD python3 healthcheck.py
```
once you clone immich and you place this on a file (e.g. Dockerfile-iml) in `immich/machine-learning`, you can CD into it and build this with:
`docker build -t immich-machine-learning:cuda50 -f Dockerfile-iml .` 

The original docker-compose should work the same.

**Reply by anhnlh** [reply]
@blackthorne can you share what you have running or the image you successfully built? I have been banging my head against the wall working on this for 2 days with no luck. I have a GTX 950M, running 580 drivers.

**Reply by blackthorne** [reply]
my setup is probably more complicated than what you need because I'm running this on a docker container inside of a LXC container on Proxmox. 

1. On the host server, I disabled nouveau drivers and installed the nvidia drivers, as follows:
```
echo 'blacklist nouveau' >> /etc/modprobe.d/blacklist.conf
echo 'options nouveau modeset=0' >> /etc/modprobe.d/blacklist.conf
sed -i 's/main/main non-free/' /etc/apt/sources.list.d/debian.sources
apt-get update
apt install -y nvidia-driver
apt install -y pve-headers-$(uname -r)
dkms install $(dkms status nvidia-current|cut -d':' -f1|cut -d',' -f1) -k $(uname -r)
reboot
# modprobe nvidia
nvidia-smi
```

   (2.) You might not need this part if you are not running it on a LXC container but then I ran [nvidia-lxc-setup.sh](https://github.com/user-attachments/files/24832399/nvidia-lxc-setup.sh)on the host machine to mount the device hardware files into the LXC container. You need to specify the right container ID.

3. I setup the system (within the LXC container) for nvidia and built the cuda50 image, as follows:
```
#!/usr/bin/bash
set -ex
apt-get update && sudo apt-get install -y --no-install-recommends \
   curl \
   gnupg2
curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
  && curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
    sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
    tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
apt update
apt install nvidia-container-toolkit -y
nvidia-ctk runtime configure --runtime=docker --set-as-default
nvidia-ctk cdi generate --output=/etc/cdi/nvidia.yaml
systemctl restart docker
git clone https://github.com/immich-app/immich.git
cd immich/machine-learning
docker build -t immich-machine-learning:cuda50 -f ../../Dockerfile-iml .
docker info | grep -i runtime
```
The content of _Dockerfile-iml_ is pasted on my previous comment.

Let me know if you have any troubles.

Francisco

**Reply by anhnlh** [reply]
I appreciate the setup. I'm mainly having trouble building the Docker image rather than the setup itself. I don't need LXC containers at the moment, so I definitely do not need the extra steps you posted.

What was the version of onnxruntime you used when building the image? I am keeping most of the Dockerfile content the same, but I had to make small adjustments for it to successfully build. I'm running the build on a more recent Windows computer (built less than a year ago), and it's quite beefy. I doubt it's a hardware requirement issue but something far more complicated than I can comprehend.

My version of the Dockerfile (.txt because GitHub is scared of extension-less files): [Dockerfile-iml.txt](https://github.com/user-attachments/files/24847796/Dockerfile-iml.txt)

Here's what I tried so far:
- Building the onnx runtime at v1.15.1 because that's the tag where the runtime was still supporting CUDA_ARCHITECTURES at 50. However, despite successfully building, the image keeps failing with various errors when run like not being able to find `ReShape`
[onnx-runtime-failure.txt](https://github.com/user-attachments/files/24847954/onnx-runtime-failure.txt).
- Consulting the ONNX compatibility table [here](https://onnxruntime.ai/docs/reference/compatibility.html), I figured I try downgrading the ONNX version used in the immich_ml image but that caused a bunch of issues not being able to find certain functions.
- Going in the opposite direction of compatibility, I also tried building onnx runtime images of higher versions. In this case, versions greater or equal to v1.18 support ONNX v1.16 (which immich_ml uses). However, my builds would never succeed, failing silently when building target `onnxruntime_providers_cuda` with exit code 2.

That's why I asked if you still have the image and possibly share it to see if I can run it off the image that's working for you.

**Reply by blackthorne** [reply]
the version of onnxruntime is the latest as per line 
`RUN git clone --recursive --depth 1 https://github.com/microsoft/onnxruntime.git onnxruntime
` in my docker file. Previously I had older versions in there but I tested the latest and worked fine. Please note that compatibility table starts with a backwards compatibility statement "Newer versions of ONNX Runtime support all models that worked with prior versions, so updates should not break integrations.".

By looking at your error, it seems like you are trying with an old version of onnx runtime.  The underlying problem could relate to your host setup hence I shared details about mine. What I'm doing in that script for LXC setup isn't really about LXC but the access to host hardware and it will be needed (one way or other) in other setups i.e. you need to make sure that the Immich container (and any virtualisation layer on top of it) has direct access to the GPU. 

I am happy to send you a tarball of the image but wouldn't be confident about it to work on your system.
